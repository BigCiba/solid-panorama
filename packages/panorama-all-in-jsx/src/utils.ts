import { NodePath } from '@babel/core';
import { addNamed } from '@babel/helper-module-imports';
import * as t from '@babel/types';

export function registerImportMethod(
    path: NodePath,
    name: string,
    moduleName: string
) {
    let imports = path.scope.getProgramParent().getData('imports');
    if (!imports) {
        imports = new Map();
        path.scope.getProgramParent().setData('imports', imports);
    }
    if (!imports.has(`${moduleName}:${name}`)) {
        let id = addNamed(path, name, moduleName, {
            nameHint: `_$${name}`
        });
        imports.set(`${moduleName}:${name}`, id);
        return id;
    } else {
        let iden = imports.get(`${moduleName}:${name}`);
        return t.cloneNode(iden);
    }
}
