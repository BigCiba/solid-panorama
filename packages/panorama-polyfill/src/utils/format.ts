// Based on https://github.com/browserify/node-util/blob/4b1c0c79790d9968eabecd2e9c786454713e200f/util.js#L33
export function format(value?: unknown, ...substitutions: unknown[]) {
    if (typeof value !== 'string') {
        return [value, ...substitutions].map(v => inspect(v)).join(' ');
    }

    let count = 0;
    let result = String(value).replace(/%[sdj%]/g, x => {
        count++;
        if (x === '%%') return '%';
        if (substitutions.length === 0) return x;

        switch (x) {
            case '%s':
                return String(substitutions.shift());

            case '%d':
                return String(Number(substitutions.shift()));

            case '%j':
                try {
                    return JSON.stringify(substitutions.shift());
                } catch {
                    return '[Circular]';
                }

            default:
                return x;
        }
    });

    for (const x of substitutions) {
        if (typeof x !== 'object' || x === null) {
            result += ` ${x}`;
        } else {
            result += ` ${inspect(x)}`;
        }
    }

    return result;
}

export function formatx(value?: unknown, ...substitutions: unknown[]) {
    if (typeof value !== 'string') {
        return [value, ...substitutions].map(v => inspect(v, true)).join(' ');
    }

    let result = String(value).replace(/%[sdj%]/g, x => {
        if (x === '%%') return '%';
        if (substitutions.length === 0) return x;

        switch (x) {
            case '%s':
                return String(substitutions.unshift());

            case '%d':
                return String(Number(substitutions.unshift()));

            case '%j':
                try {
                    return JSON.stringify(substitutions.unshift());
                } catch {
                    return '[Circular]';
                }

            default:
                return x;
        }
    });

    for (const x of substitutions) {
        if (typeof x !== 'object' || x === null) {
            result += ` ${x}`;
        } else {
            result += ` ${inspect(x)}`;
        }
    }

    return result;
}

function inspect(value: unknown, expand?: boolean, tab = '') {
    let result = '';
    if (typeof value === 'string') {
        result = `"${value}"`;
    } else if (typeof value === 'number' || typeof value === 'boolean') {
        result = `${value}`;
    } else if (typeof value === 'function') {
        result = isClass(value)
            ? `[class ${value.name}]`
            : `[function ${value.name}]`;
    } else if (typeof value === 'symbol') {
        result = value.toString();
    } else if (typeof value === 'undefined') {
        result = `undefined`;
    } else if (typeof value === 'bigint') {
        result = `[bigint ${value.toString()}]`;
    } else if (typeof value === 'object') {
        if (value === null) {
            result = 'null';
        } else if (Array.isArray(value)) {
            let list = [];
            for (const v of value) {
                list.push(
                    tab + inspect(v, expand, expand ? tab + '    ' : tab)
                );
            }
            if (expand) {
                result += '[\n';
                result += list.map(v => '    ' + v).join(',\n');
                result += '\n' + tab + ']';
            } else {
                result = `[ ${list.join(', ')} ]`;
            }
        } else {
            let list: string[] = [];
            let prefix = '';
            // Map
            if (value instanceof Map) {
                prefix = '[Map]';
                for (const [k, v] of value.entries()) {
                    let key = '';
                    if (typeof k === 'object') {
                        key = Array.isArray(k) ? '[Array]' : '[Object]';
                    } else if (k.toString) {
                        key = k.toString();
                    }
                    list.push(
                        `${tab}${key}: ${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            } else if (value instanceof Set) {
                prefix = '[Set]';
                for (const v of value.values()) {
                    list.push(
                        `${tab}${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            } else {
                const isPanel = isPanelBase(value);
                for (const [k, v] of Object.entries(value)) {
                    if (isPanel) {
                        console.log(value, k, v);
                    }
                    if (k === 'style' && isPanel) {
                        list.push(`${tab}${k}: [VCSSStyleDeclaration]`);
                        continue;
                    }
                    list.push(
                        `${tab}${k}: ${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            }

            if (expand) {
                result += prefix + '{\n';
                result += list.map(v => '    ' + v).join(',\n');
                result += '\n' + tab + '}';
            } else {
                result = prefix + `{ ${list.join(', ')} }`;
            }
        }
    }
    return result;
}

const isPanelBase = (value: object): value is PanelBase =>
    'paneltype' in value &&
    'rememberchildfocus' in value &&
    'SetPanelEvent' in value;

function isClass(value: unknown) {
    if (typeof value !== 'function') return false;
    const descriptor = Object.getOwnPropertyDescriptor(value, 'prototype');
    if (!descriptor) return false;
    return !descriptor.writable;
}
