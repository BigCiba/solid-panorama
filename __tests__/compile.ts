import { describe, expect, test } from '@jest/globals';
import { parser } from './utils';

describe('compile', function () {
    test('transform: normal', function () {
        const result = parser(`
            import { render } from 'solid-panorama-runtime';

            function Item(props) {
                const [rootStyle, setRootStyle] = createSignal('root');
                let visible = true
                let root, btnA
                return (
                    <Panel id="root" ref={root} class={rootStyle().join(' ')}>
                        <Label text="Testing" />
                        <Panel class="buttons">
                            <Button ref={(a) => btnA = a} />
                            <Button visible={visible} />
                            <Button />
                        </Panel>
                    </Panel>
                );
            }
            
            function HelloWorld() {
                return (
                    <Panel>
                        <Item />
                        <Item show />
                    </Panel>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));        
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: fragment', function () {
        expect(() => {
            parser(`
                import { render } from 'solid-panorama-runtime';
                
                function HelloWorld() {
                    return (
                        <>
                            <Button />
                            <Button />
                            <Button />
                        </>
                    );
                }
                
                render(() => <HelloWorld />, $('#app'));        
            `);
        }).toThrow();
    });

    test('transform: for each', function () {
        const result = parser(`
            import { render } from "solid-js/web";
            import { createSignal } from "solid-js";
            
            const App = () => {
                const [list, setList] = createSignal(['A']);
                
                function click() {
                    setList([...list(), "B"])
                }
                
                return <Panel>
                    <For each={list()} >
                        {(item, index) => {
                        return <Label text={item} />
                        }}
                    </For>
                    <Button onClick={click}>Click</Button>
                </Panel>;
            };
            
            render(() => <App />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: style', function () {
        const result = parser(`
            import { render } from 'solid-panorama-runtime';
            
            function HelloWorld() {
                return (
                    <Button style={{width:'12px', height: 12}} />
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: textNode', function () {
        const result = parser(`
            import { render } from 'solid-panorama-runtime';
            
            function HelloWorld() {
                return (
                    <Panel>
                        <Panel>
                            Welcome My Game
                        </Panel>
                        <Panel>
                            {\`<strong>Welcome</strong> My Game\`}
                        </Panel>
                        <Panel>
                            <Label text="Welcome" />
                            #addon_game_name
                            <Label text="(～￣▽￣)～" />
                        </Panel>
                    </Panel>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });
});
