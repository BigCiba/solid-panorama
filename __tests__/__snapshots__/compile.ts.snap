// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile transform: for each 1`] = `
"import { insert as _$insert } from "solid-panorama-runtime";
import { createTextNode as _$createTextNode } from "solid-panorama-runtime";
import { setProp as _$setProp } from "solid-panorama-runtime";
import { createComponent as _$createComponent } from "solid-panorama-runtime";
import { For as _$For } from "solid-panorama-runtime";
import { createElement as _$createElement } from "solid-panorama-runtime";
import { render } from "solid-js/web";
import { createSignal } from "solid-js";
const App = () => {
  const [list, setList] = createSignal(['A']);
  function click() {
    setList([...list(), "B"]);
  }
  return (() => {
    const _el$ = _$createElement("Panel", {}, null),
      _el$2 = _$createElement("Button", {
        onClick: click
      }, _el$);
    _$insert(_el$, _$createComponent(_$For, {
      get each() {
        return list();
      },
      children: (item, index) => {
        return (() => {
          const _el$4 = _$createElement("Label", {
            text: item
          }, null);
          _$setProp(_el$4, "text", item);
          return _el$4;
        })();
      }
    }, _el$), _el$2);
    _$createTextNode(\`Click\`, _el$2);
    _$setProp(_el$2, "onClick", click);
    return _el$;
  })();
};
render(() => _$createComponent(App, {}, null), $('#app'));"
`;

exports[`compile transform: normal 1`] = `
"import { insert as _$insert } from "solid-panorama-runtime";
import { createComponent as _$createComponent } from "solid-panorama-runtime";
import { effect as _$effect } from "solid-panorama-runtime";
import { setProp as _$setProp } from "solid-panorama-runtime";
import { use as _$use } from "solid-panorama-runtime";
import { createElement as _$createElement } from "solid-panorama-runtime";
import { render } from 'solid-panorama-runtime';
function Item(props) {
  const [rootStyle, setRootStyle] = createSignal('root');
  let visible = true;
  let root, btnA;
  return (() => {
    const _el$ = _$createElement("Panel", {
        id: "root"
      }, null),
      _el$2 = _$createElement("Label", {
        text: "Testing"
      }, _el$),
      _el$3 = _$createElement("Panel", {
        className: "buttons"
      }, _el$),
      _el$4 = _$createElement("Button", {}, _el$3),
      _el$5 = _$createElement("Button", {
        visible: visible
      }, _el$3),
      _el$6 = _$createElement("Button", {}, _el$3);
    const _ref$ = root;
    typeof _ref$ === "function" ? _$use(_ref$, _el$) : root = _el$;
    _$use(a => btnA = a, _el$4);
    _$setProp(_el$5, "visible", visible);
    _$effect(_$p => _$setProp(_el$, "className", rootStyle().join(' '), _$p));
    return _el$;
  })();
}
function HelloWorld() {
  return (() => {
    const _el$7 = _$createElement("Panel", {}, null);
    _$insert(_el$7, _$createComponent(Item, {}, _el$7), null);
    _$insert(_el$7, _$createComponent(Item, {
      show: true
    }, _el$7), null);
    return _el$7;
  })();
}
render(() => _$createComponent(HelloWorld, {}, null), $('#app'));"
`;

exports[`compile transform: style 1`] = `
"import { createComponent as _$createComponent } from "solid-panorama-runtime";
import { setProp as _$setProp } from "solid-panorama-runtime";
import { createElement as _$createElement } from "solid-panorama-runtime";
import { render } from 'solid-panorama-runtime';
function HelloWorld() {
  return (() => {
    const _el$ = _$createElement("Button", {}, null);
    _$setProp(_el$, "style", {
      width: '12px',
      height: 12
    });
    return _el$;
  })();
}
render(() => _$createComponent(HelloWorld, {}, null), $('#app'));"
`;

exports[`compile transform: textNode 1`] = `
"import { createComponent as _$createComponent } from "solid-panorama-runtime";
import { createTextNode as _$createTextNode } from "solid-panorama-runtime";
import { createElement as _$createElement } from "solid-panorama-runtime";
import { render } from 'solid-panorama-runtime';
function HelloWorld() {
  return (() => {
    const _el$ = _$createElement("Panel", {}, null),
      _el$2 = _$createElement("Panel", {}, _el$),
      _el$4 = _$createElement("Panel", {}, _el$),
      _el$6 = _$createElement("Panel", {}, _el$),
      _el$7 = _$createElement("Label", {
        text: "Welcome"
      }, _el$6),
      _el$8 = _$createTextNode(\`#addon_game_name\`, _el$6),
      _el$9 = _$createElement("Label", {
        text: "(～￣▽￣)～"
      }, _el$6);
    _$createTextNode(\`Welcome My Game\`, _el$2);
    _$createTextNode(\`<strong>Welcome</strong> My Game\`, _el$4);
    return _el$;
  })();
}
render(() => _$createComponent(HelloWorld, {}, null), $('#app'));"
`;
